### ISSUE TITLE
Patterns are created multiple times

### ISSUE THREAD

**Issue by erhanleblebici** [eMoflon/emoflon-ibex#217]
Something seems to be wrong with the pattern factory.
In the attached screenshot, you see a Map<Rule, Collecion<IbexPattern>> which is the result of TGGCompiler.

I have three rules (black rectangles) but six Src-Patterns.
And it is not deterministic which patterns are created multiple times. (I had duplicate NO_FILTER_AC_SRC in the next run)

<img width="969" alt="screenshot 2017-07-07 18 18 10" src="https://user-images.githubusercontent.com/11506522/27967419-490bdeba-6343-11e7-9283-35793a5e6664.png">

**Comment by anthonyanjorin** [comment]
@FStolte Could this be one of the problems we solved already?  Perhaps you can remember better than I can.

**Comment by FStolte** [comment]
I don't recall that we encountered a situation where duplicate patterns were created non-deterministically. For all bugs we already fixed, the creation of duplicate patterns (or duplicate invocations) was a systematic error.

Besides, it seems rather strange to me that Src-Patterns can be duplicated at all with how the Pattern Factory works. Since the patterns are stored in a map with their classes as key, it should be impossible to store two Src-Patterns for the same rule, even if they would be created twice erroneously.
